kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/eq/DataClassEq.kt

```diff
@@ -0,0 +1,87 @@
+package io.kotest.assertions.eq
+
+import io.kotest.assertions.Actual
+import io.kotest.assertions.Expected
+import io.kotest.assertions.failure
+import io.kotest.assertions.show.show
+import kotlin.reflect.KCallable
+import kotlin.reflect.KClass
+import kotlin.reflect.KProperty1
+
+private val <T : Any> KClass<T>.dataClassProperties: List<KCallable<*>>
+   get() {
+      val dataClassConstructorParams = ::constructors.get().last().parameters
+      val membersByName = ::members.get().associateBy(KCallable<*>::name)
+      return dataClassConstructorParams.map { membersByName[it.name] }.filterNotNull()
+   }
+
+/**
+ * Prints a detailed diff of data class instances.
+ *
+ * HELP! Where should this class live, it contains reflection code and should only be reachable if the platform
+ * supports data class diff (See Eq.kt)
+ */
+internal object DataClassEq : Eq<Any> {
```

Comment:
All reflection code should live in kotest-commons in an existing `interface Reflection`.